File: adapters/data/in-memory-store.ts
// adapters/data/in-memory-store.ts
// InMemoryDataStore — reference IDataStore implementation for testing.
// Deterministic, zero-dependency, no I/O. Used in unit tests and local dev.
// Also serves as the conformance reference for all other IDataStore adapters.

import { randomUUID } from 'node:crypto'
import type {
  CollectionSchema,
  DataStoreKind,
  IDataStore,
  ITransaction,
  IsolationLevel,
  PagedResult,
  QueryOptions,
  WhereClause,
} from './protocols.js'

class InMemoryTransaction implements ITransaction {
  private _active = true
  private readonly _snapshots: Map<string, Map<string, Record<string, unknown>>>
  private readonly _store: InMemoryDataStore

  constructor(store: InMemoryDataStore, snapshots: Map<string, Map<string, Record<string, unknown>>>) {
    this._store = store
    this._snapshots = snapshots
  }

  get isActive(): boolean { return this._active }

  async commit(): Promise<void> {
    if (!this._active) throw new Error('Transaction already completed')
    this._active = false
    // Changes are already applied to the live store; commit is a no-op here
    // (optimistic concurrency — real adapters implement MVCC)
  }

  async rollback(): Promise<void> {
    if (!this._active) throw new Error('Transaction already completed')
    this._active = false
    // Restore pre-transaction snapshots
    this._store._restoreSnapshots(this._snapshots)
  }
}

export class InMemoryDataStore implements IDataStore {
  readonly kind:    DataStoreKind = 'nosql'
  readonly storeId: string

  private _connected = false
  private readonly _collections = new Map<string, Map<string, Record<string, unknown>>>()
  private readonly _schemas     = new Map<string, CollectionSchema>()

  constructor(storeId = 'in-memory') {
    this.storeId = storeId
  }

  get isConnected(): boolean { return this._connected }

  async connect():    Promise<void> { this._connected = true }
  async disconnect(): Promise<void> { this._connected = false }
  async ping():       Promise<boolean> { return this._connected }

  async createCollection(schema: CollectionSchema): Promise<void> {
    if (!this._collections.has(schema.name)) {
      this._collections.set(schema.name, new Map())
    }
    this._schemas.set(schema.name, schema)
  }

  async dropCollection(name: string): Promise<void> {
    this._collections.delete(name)
    this._schemas.delete(name)
  }

  async collectionExists(name: string): Promise<boolean> {
    return this._collections.has(name)
  }

  async listCollections(): Promise<string[]> {
    return Array.from(this._collections.keys())
  }

  async insert<T extends Record<string, unknown>>(collection: string, record: T): Promise<string> {
    const coll = this._getCollection(collection)
    const id   = (record['id'] as string | undefined) ?? randomUUID()
    coll.set(id, { ...record, id })
    return id
  }

  async insertMany<T extends Record<string, unknown>>(collection: string, records: T[]): Promise<string[]> {
    return Promise.all(records.map(r => this.insert(collection, r)))
  }

  async findById<T extends Record<string, unknown>>(collection: string, id: string): Promise<T | null> {
    const coll = this._getCollection(collection)
    return (coll.get(id) as T | undefined) ?? null
  }

  async find<T extends Record<string, unknown>>(
    collection: string,
    options:    QueryOptions<T> = {},
  ): Promise<PagedResult<T>> {
    const coll   = this._getCollection(collection)
    let   results = Array.from(coll.values()) as T[]

    if (options.where) {
      results = results.filter(r => this._applyWhere(r, options.where!))
    }

    if (options.orderBy) {
      for (const { field, direction } of [...options.orderBy].reverse()) {
        results.sort((a, b) => {
          const av = a[field as string] as number | string
          const bv = b[field as string] as number | string
          const cmp = av < bv ? -1 : av > bv ? 1 : 0
          return direction === 'asc' ? cmp : -cmp
        })
      }
    }

    const total   = results.length
    const offset  = options.offset  ?? 0
    const limit   = options.limit   ?? 100
    const page    = results.slice(offset, offset + limit)

    if (options.select) {
      const keys = options.select as string[]
      return {
        data:    page.map(r => Object.fromEntries(keys.map(k => [k, r[k]])) as T),
        total, limit, offset,
        hasMore: offset + limit < total,
      }
    }

    return { data: page, total, limit, offset, hasMore: offset + limit < total }
  }

  async findOne<T extends Record<string, unknown>>(
    collection: string,
    where:      WhereClause[],
  ): Promise<T | null> {
    const result = await this.find<T>(collection, { where, limit: 1 })
    return result.data[0] ?? null
  }

  async update<T extends Record<string, unknown>>(
    collection: string,
    id:         string,
    patch:      Partial<T>,
  ): Promise<boolean> {
    const coll = this._getCollection(collection)
    const existing = coll.get(id)
    if (!existing) return false
    coll.set(id, { ...existing, ...patch, id })
    return true
  }

  async upsert<T extends Record<string, unknown>>(
    collection:  string,
    record:      T,
    matchFields: (keyof T)[],
  ): Promise<{ id: string; created: boolean }> {
    const where: WhereClause[] = matchFields.map(f => ({
      field:    f as string,
      operator: '=' as const,
      value:    record[f],
    }))
    const existing = await this.findOne<T>(collection, where)
    if (existing) {
      const id = existing['id'] as string
      await this.update(collection, id, record)
      return { id, created: false }
    }
    const id = await this.insert(collection, record)
    return { id, created: true }
  }

  async delete(collection: string, id: string): Promise<boolean> {
    const coll = this._getCollection(collection)
    return coll.delete(id)
  }

  async count(collection: string, where?: WhereClause[]): Promise<number> {
    const result = await this.find(collection, { where, limit: Number.MAX_SAFE_INTEGER })
    return result.total
  }

  async beginTransaction(_isolation?: IsolationLevel): Promise<ITransaction> {
    // Snapshot all collections for rollback
    const snapshots = new Map<string, Map<string, Record<string, unknown>>>()
    for (const [name, coll] of this._collections) {
      snapshots.set(name, new Map(Array.from(coll.entries()).map(([k, v]) => [k, { ...v }])))
    }
    return new InMemoryTransaction(this, snapshots)
  }

  async bulkInsert<T extends Record<string, unknown>>(
    collection: string,
    records:    T[],
    options:    { batchSize?: number; onError?: 'abort' | 'skip' } = {},
  ): Promise<{ inserted: number; failed: number; errors: Array<{ index: number; error: string }> }> {
    const onError = options.onError ?? 'abort'
    let inserted = 0
    let failed   = 0
    const errors: Array<{ index: number; error: string }> = []

    for (let i = 0; i < records.length; i++) {
      try {
        await this.insert(collection, records[i])
        inserted++
      } catch (err) {
        failed++
        errors.push({ index: i, error: (err as Error).message })
        if (onError === 'abort') break
      }
    }

    return { inserted, failed, errors }
  }

  // Internal: used by InMemoryTransaction.rollback()
  _restoreSnapshots(snapshots: Map<string, Map<string, Record<string, unknown>>>): void {
    for (const [name, snapshot] of snapshots) {
      this._collections.set(name, snapshot)
    }
  }

  private _getCollection(name: string): Map<string, Record<string, unknown>> {
    if (!this._collections.has(name)) {
      this._collections.set(name, new Map())
    }
    return this._collections.get(name)!
  }

  private _applyWhere(record: Record<string, unknown>, clauses: WhereClause[]): boolean {
    return clauses.every(({ field, operator, value }) => {
      const rv = record[field]
      switch (operator) {
        case '=':       return rv === value
        case '!=':      return rv !== value
        case '<':       return (rv as number) < (value as number)
        case '<=':      return (rv as number) <= (value as number)
        case '>':       return (rv as number) > (value as number)
        case '>=':      return (rv as number) >= (value as number)
        case 'in':      return Array.isArray(value) && value.includes(rv)
        case 'not_in':  return Array.isArray(value) && !value.includes(rv)
        case 'like':    return typeof rv === 'string' && rv.includes(String(value).replace(/%/g, ''))
        case 'exists':  return rv !== undefined && rv !== null
        default:        return true
      }
    })
  }
}

File: src/ooagent/adapters/data/in_memory_store.py
"""adapters/data/in_memory_store.py — InMemoryDataStore.

Reference IDataStore implementation for testing. Deterministic,
zero-dependency, no I/O. Used in unit tests and local dev. Also serves as the
conformance reference for all other IDataStore adapters.
"""

from __future__ import annotations

import functools
import uuid
from collections.abc import Callable
from typing import Any, Literal

from ooagent.adapters.data.protocols import (
    CollectionSchema,
    DataStoreKind,
    IDataStore,
    IsolationLevel,
    ITransaction,
    PagedResult,
    QueryOptions,
    SortOrder,
    TransactionError,
    WhereClause,
)

# Number.MAX_SAFE_INTEGER — used by count() to page through every record.
_MAX_SAFE_INTEGER = 2**53 - 1


def _make_comparator(
    field: str, direction: SortOrder
) -> Callable[[dict[str, Any], dict[str, Any]], int]:
    """Builds a JS-`Array.prototype.sort`-equivalent 2-arg comparator.

    Values that are not directly ordering-comparable (e.g. mixed types) are
    treated as equal rather than raising — mirroring JS's permissive `<`/`>`
    operators, which never throw for this comparison.
    """

    def _cmp(a: dict[str, Any], b: dict[str, Any]) -> int:
        av = a.get(field)
        bv = b.get(field)
        try:
            # av/bv may be None or otherwise mutually non-comparable (e.g.
            # str vs int); the surrounding try/except TypeError is the
            # intentional guard for that (mirrors JS's permissive `<`/`>`,
            # which never throws) — not a bug to fix here.
            if av < bv:  # type: ignore[operator]
                cmp = -1
            elif av > bv:  # type: ignore[operator]
                cmp = 1
            else:
                cmp = 0
        except TypeError:
            cmp = 0
        return cmp if direction == "asc" else -cmp

    return _cmp


class InMemoryTransaction(ITransaction):
    def __init__(
        self,
        store: InMemoryDataStore,
        snapshots: dict[str, dict[str, dict[str, Any]]],
    ) -> None:
        self._active = True
        self._snapshots = snapshots
        self._store = store

    @property
    def is_active(self) -> bool:
        return self._active

    async def commit(self) -> None:
        if not self._active:
            raise TransactionError("Transaction already completed")
        self._active = False
        # Changes are already applied to the live store; commit is a no-op
        # here (optimistic concurrency — real adapters implement MVCC).

    async def rollback(self) -> None:
        if not self._active:
            raise TransactionError("Transaction already completed")
        self._active = False
        # Restore pre-transaction snapshots.
        self._store._restore_snapshots(self._snapshots)


class InMemoryDataStore(IDataStore):
    def __init__(self, store_id: str = "in-memory") -> None:
        self._store_id = store_id
        self._connected = False
        self._collections: dict[str, dict[str, dict[str, Any]]] = {}
        self._schemas: dict[str, CollectionSchema] = {}

    @property
    def kind(self) -> DataStoreKind:
        return "nosql"

    @property
    def store_id(self) -> str:
        return self._store_id

    @property
    def is_connected(self) -> bool:
        return self._connected

    async def connect(self) -> None:
        self._connected = True

    async def disconnect(self) -> None:
        self._connected = False

    async def ping(self) -> bool:
        return self._connected

    async def create_collection(self, schema: CollectionSchema) -> None:
        if schema.name not in self._collections:
            self._collections[schema.name] = {}
        self._schemas[schema.name] = schema

    async def drop_collection(self, name: str) -> None:
        self._collections.pop(name, None)
        self._schemas.pop(name, None)

    async def collection_exists(self, name: str) -> bool:
        return name in self._collections

    async def list_collections(self) -> list[str]:
        return list(self._collections.keys())

    async def insert(self, collection: str, record: dict[str, Any]) -> str:
        coll = self._get_collection(collection)
        record_id = record.get("id")
        id_ = record_id if record_id is not None else str(uuid.uuid4())
        coll[id_] = {**record, "id": id_}
        return id_

    async def insert_many(self, collection: str, records: list[dict[str, Any]]) -> list[str]:
        ids = []
        for r in records:
            ids.append(await self.insert(collection, r))
        return ids

    async def find_by_id(self, collection: str, id: str) -> dict[str, Any] | None:
        coll = self._get_collection(collection)
        return coll.get(id)

    async def find(
        self, collection: str, options: QueryOptions | None = None
    ) -> PagedResult[dict[str, Any]]:
        options = options or QueryOptions()
        coll = self._get_collection(collection)
        results: list[dict[str, Any]] = list(coll.values())

        if options.where:
            results = [r for r in results if self._apply_where(r, options.where)]

        if options.order_by:
            for spec in reversed(options.order_by):
                results.sort(key=functools.cmp_to_key(_make_comparator(spec.field, spec.direction)))

        total = len(results)
        offset = options.offset if options.offset is not None else 0
        limit = options.limit if options.limit is not None else 100
        page = results[offset : offset + limit]

        if options.select:
            keys = options.select
            data = [{k: r.get(k) for k in keys} for r in page]
            return PagedResult(
                data=data,
                total=total,
                limit=limit,
                offset=offset,
                has_more=offset + limit < total,
            )

        return PagedResult(
            data=page,
            total=total,
            limit=limit,
            offset=offset,
            has_more=offset + limit < total,
        )

    async def find_one(self, collection: str, where: list[WhereClause]) -> dict[str, Any] | None:
        result = await self.find(collection, QueryOptions(where=where, limit=1))
        return result.data[0] if result.data else None

    async def update(self, collection: str, id: str, patch: dict[str, Any]) -> bool:
        coll = self._get_collection(collection)
        existing = coll.get(id)
        if existing is None:
            return False
        coll[id] = {**existing, **patch, "id": id}
        return True

    async def upsert(
        self, collection: str, record: dict[str, Any], match_fields: list[str]
    ) -> dict[str, Any]:
        where = [WhereClause(field=f, operator="=", value=record.get(f)) for f in match_fields]
        existing = await self.find_one(collection, where)
        if existing is not None:
            id_ = existing["id"]
            await self.update(collection, id_, record)
            return {"id": id_, "created": False}
        id_ = await self.insert(collection, record)
        return {"id": id_, "created": True}

    async def delete(self, collection: str, id: str) -> bool:
        coll = self._get_collection(collection)
        return coll.pop(id, None) is not None

    async def count(self, collection: str, where: list[WhereClause] | None = None) -> int:
        result = await self.find(collection, QueryOptions(where=where, limit=_MAX_SAFE_INTEGER))
        return result.total

    async def begin_transaction(self, isolation: IsolationLevel | None = None) -> ITransaction:
        # Snapshot all collections for rollback.
        snapshots: dict[str, dict[str, dict[str, Any]]] = {
            name: {k: {**v} for k, v in coll.items()} for name, coll in self._collections.items()
        }
        return InMemoryTransaction(self, snapshots)

    async def bulk_insert(
        self,
        collection: str,
        records: list[dict[str, Any]],
        batch_size: int | None = None,
        on_error: Literal["abort", "skip"] = "abort",
    ) -> dict[str, Any]:
        inserted = 0
        failed = 0
        errors: list[dict[str, Any]] = []

        for i, record in enumerate(records):
            try:
                await self.insert(collection, record)
                inserted += 1
            except Exception as err:  # noqa: BLE001 - mirrors TS catch-all
                failed += 1
                errors.append({"index": i, "error": str(err)})
                if on_error == "abort":
                    break

        return {"inserted": inserted, "failed": failed, "errors": errors}

    # Internal: used by InMemoryTransaction.rollback()
    def _restore_snapshots(self, snapshots: dict[str, dict[str, dict[str, Any]]]) -> None:
        for name, snapshot in snapshots.items():
            self._collections[name] = snapshot

    def _get_collection(self, name: str) -> dict[str, dict[str, Any]]:
        if name not in self._collections:
            self._collections[name] = {}
        return self._collections[name]

    def _apply_where(self, record: dict[str, Any], clauses: list[WhereClause]) -> bool:
        def _matches(clause: WhereClause) -> bool:
            rv = record.get(clause.field)
            value = clause.value
            op = clause.operator
            if op == "=":
                return bool(rv == value)
            if op == "!=":
                return bool(rv != value)
            if op == "<":
                return bool(rv < value)
            if op == "<=":
                return bool(rv <= value)
            if op == ">":
                return bool(rv > value)
            if op == ">=":
                return bool(rv >= value)
            if op == "in":
                return isinstance(value, list) and rv in value
            if op == "not_in":
                return isinstance(value, list) and rv not in value
            if op == "like":
                return isinstance(rv, str) and str(value).replace("%", "") in rv
            if op == "exists":
                return rv is not None
            return True

        return all(_matches(c) for c in clauses)
